### PR TITLE
CASSANDRA-17663 Ensure FileStreamTask cannot compromise shared channel proxy for system table when interrupted

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-17663

### CODE REVIEW

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-17663-4.1&filter=all

File: src/java/org/apache/cassandra/streaming/StreamingChannel.java [hunk 1]
```diff
@@ -49,6 +49,14 @@ public static void unsafeSet(StreamingChannel.Factory factory)
         }
 
         StreamingChannel create(InetSocketAddress to, int messagingVersion, Kind kind) throws IOException;
+
+        default StreamingChannel create(InetSocketAddress to,
+                                        InetSocketAddress preferred,
+                                        int messagingVersion,
+                                        StreamingChannel.Kind kind) throws IOException
+        {
+            throw new UnsupportedOperationException();
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/streaming/StreamingChannel.java]: so fail if `BulkLoadConnectionFactory`?

File: src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java [hunk 2]
```diff
@@ -202,7 +210,9 @@ public Future<?> sendMessage(StreamingChannel channel, StreamMessage message)
                 throw new RuntimeException("Cannot send stream data messages for preview streaming sessions");
             if (logger.isDebugEnabled())
                 logger.debug("{} Sending {}", createLogTag(session), message);
-            return fileTransferExecutor.submit(new FileStreamTask((OutgoingStreamMessage)message));
+
+            InetAddressAndPort connectTo = SystemKeyspace.getPreferredIP(to);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java:214]: do we know If there are any paths where this is called on a Netty thread?

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java [hunk 3]
```diff
@@ -124,18 +128,58 @@ public void testRemoteSyncFailure() throws Exception
...
                                            .describedAs("node%d logged hidden exception org.apache.cassandra.repair.SomeRepairFailedException", i.config().num())
                                            .isEmpty());
         }
     }
 
+    @Test
+    public void testRemoteStreamFailure() throws Exception
+    {
+        try (Cluster cluster = init(Cluster.build(3)
+                                           .withConfig(config -> config.with(GOSSIP)
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java]: nit, `GOSSIP, NETWORK`

**Review by dcapwell — COMMENTED** [review]
getting pulled away but will look closer later

File: src/java/org/apache/cassandra/streaming/StreamingChannel.java (hunk 1 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/streaming/StreamingChannel.java]: `BulkLoadConnectionFactory extends NettyStreamingConnectionFactory` right?

File: src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java (hunk 2 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java:214]: `StreamSession#startStreamingFiles()` is the only thing that can call `sendMessage()` here with a `OutgoingStreamMessage`. I don't think we can hit that from an even loop thread, but I'll do a quick audit...

File: test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java (hunk 3 above)
Comment (maedhroz) [reply test/distributed/org/apache/cassandra/distributed/test/RepairErrorsTest.java]: ah, TIL

File: src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java (hunk 2 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java:214]: Yeah, this only happens in `Stream-Deserializer` threads running `StreamDeserializingTask`s (responding to `PREPARE_SYNACK` and `PREPARE_ACK`).

File: src/java/org/apache/cassandra/streaming/StreamingChannel.java (hunk 1 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/streaming/StreamingChannel.java]: Ah, but we still need to provide differentiated behavior in the `BulkLoadConnectionFactory` version or `create()`.

File: src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java [hunk 4]
```diff
@@ -155,21 +155,29 @@ private void setupControlMessageChannel() throws IOException
...
-                                          new StreamDeserializingTask(session, channel, messagingVersion));
-            session.attachInbound(channel);
-        }
+        StreamingChannel channel = factory.create(to, messagingVersion, StreamingChannel.Kind.CONTROL);
+        executorFactory().startThread(String.format("Stream-Deserializer-%s-%s", to.toString(), channel.id()),
+                                      new StreamDeserializingTask(session, channel, messagingVersion));
+        session.attachInbound(channel);
+        session.attachOutbound(channel);
+
+        logger.debug("Creating {}", channel.description());
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java]: As we know the difference, might be kind to log `Creating control {}`

File: src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java [hunk 5]
```diff
@@ -155,21 +155,29 @@ private void setupControlMessageChannel() throws IOException
...
+    }
+    
+    private StreamingChannel createFileChannel(InetAddressAndPort connectTo) throws IOException
+    {
+        logger.debug("Creating stream session to {} as {}", to, session.isFollower() ? "follower" : "initiator");
+
+        StreamingChannel channel = factory.create(to, connectTo, messagingVersion, StreamingChannel.Kind.FILE);
         session.attachOutbound(channel);
 
         logger.debug("Creating {}", channel.description());
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/streaming/async/StreamingMultiplexedChannel.java]: Similarly `Creating file {}`

File: src/java/org/apache/cassandra/streaming/StreamingChannel.java (hunk 1 above)
Comment (jonmeredith) [reply src/java/org/apache/cassandra/streaming/StreamingChannel.java]: Why provide the default interface method? Are there cases where you wouldn't implement this?

**Review by jonmeredith — APPROVED** [review]
Looks good to me - just a couple of nits/questions.

File: src/java/org/apache/cassandra/streaming/StreamingChannel.java (hunk 1 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/streaming/StreamingChannel.java]: Check the latest commit. There probably would have been an issue for the channel factory we create in `DirectStreamingConnectionFactory`. The default behavior of ignoring the "preferred" address is, I think, what we want for the simulator (which creates direct channels) and the other test usage (which just want to blow up in someone calls `create()`).

**Review by dcapwell — APPROVED** [review]

**Review by jonmeredith — APPROVED** [review]

**Comment by maedhroz** [comment]
Committed in 663e51d92868cbd045a83d7fa53e373bb28721a0

**Pull request closed without merging**